### ISSUE TITLE
"Quiet" output? A.k.a: where is this visible return char (⏎) coming from?

### ISSUE THREAD

**Issue by beporter** [bobthecow/psysh#211]
Sample transcript:

``` shell
$ mkdir psytest; cd psytest/

$ composer init
# ...snip...

$ composer require psy/psysh:@stable
./composer.json has been updated
Loading composer repositories with package information
Updating dependencies (including require-dev)
  - Installing jakub-onderka/php-console-color (0.1)
    Downloading: 100%         

  - Installing jakub-onderka/php-console-highlighter (v0.3.2)
    Downloading: 100%         

  - Installing dnoegel/php-xdg-base-dir (0.1)
    Downloading: 100%         

  - Installing nikic/php-parser (v1.3.0)
    Downloading: 100%         

  - Installing symfony/console (v2.7.1)
    Downloading: 100%         

  - Installing psy/psysh (v0.4.4)
    Downloading: 100%         
# ...snip...

$ echo 'echo serialize(["alpha", "beta", 3]); exit;' | vendor/bin/psysh 
Psy Shell v0.4.4 (PHP 5.6.10 — cli) by Justin Hileman
a:3:{i:0;s:5:"alpha";i:1;s:4:"beta";i:2;i:3;}⏎
```

I have a request and a question: 
1. Because I'm trying to capture output, it would be nice to be able to suppress the "banner" line that display's the Psy Shell version using a `-q` or `--quiet` command line option.
2. Where is that visible "return" character (⏎) at the very end coming from? I can't seem to find it [in Psysh's source](https://github.com/bobthecow/psysh/search?utf8=%E2%9C%93&q=%E2%8F%8E), nor [in Symfony/Console's source](https://github.com/symfony/Console/search?utf8=%E2%9C%93&q=%E2%8F%8E). Additionally, why does `serialize()` (and `var_export()` produce this character, but `print_r()` and `var_dump()` do not?

Disclaimers:
- Yes, this is a contrived example-- pretend the output is something generated from a larger app. I can provide that example if it's difficult to imagine the use case.
- Yes, I realize I can use `vendor/bin/psysh | tail -n+2 | sed s/⏎//`. That's silly to have to do every time and as an answer doesn't help anyone understand what the code is doing.

**Comment by bobthecow** [comment]
Yeah, a `--quiet` option would prolly make sense, though I'm not sure how much use a REPL has from outside the REPL :)

The visible "return" character is an idea stolen from the Fish shell. Try this in `php -a` and `psysh` (and any other PHP REPLs you might have handy):

```
>>> echo "wat";
>>> echo "wat\n";
```

You have four options, and most REPLs and shells pick the wrong one:

1. You can never add a newline to output. This means if you `echo "foo";` the prompt for your next line gets mashed into the output from the current. This is what `php -a` used to do, and it's pretty ugly.

2. You can always add a newline to output. This means if you `echo "bar";` everything is fine, and the next line's prompt isn't in a weird place, but if you `echo "baz\n";` then you get an ugly extra newline after your output and before the next line's prompt. There are several PHP shells and REPLs, including Bash, which choose this path. It's less ugly, but still ugly.

3. You can detect whether the output ends in a newline, and add one yourself if it doesn't. This means your prompt is never mashed onto the same line as the output, and it's never hanging out two lines below. Unfortunately, it also makes it impossible to tell whether your output ended in a newline or not. This is what PHP currently does.

4. You can detect whether the output ends in a newline, and add one yourself if it doesn't, but let the user know what you did. That's what Fish and PsySH do: the `⏎` character is added any time an implicit newline had to be added to the output to keep things sane. That's why it's in a different color as well.

You can see it implemented here: https://github.com/bobthecow/psysh/blob/master/src/Psy/Shell.php#L599-L625

**Comment by beporter** [comment]
Thank you for the awesome explanation, that's extremely helpful and makes perfect sense, of course. (Weird that Github search doesn't return a match for that line from `writeStdout()` but whatever.)

So basically, if I want to suppress the Psysh-provided implicit visible return, I need to just make sure I provide my own. So comparing before and after:

```shell
$ echo 'echo "implicit (visible) return will be added"; exit;' | vendor/bin/psysh 
Psy Shell v0.4.4 (PHP 5.6.10 — cli) by Justin Hileman
implicit (visible) return will be added⏎
$ echo 'echo "providing my own\n"; exit;' | vendor/bin/psysh 
Psy Shell v0.4.4 (PHP 5.6.10 — cli) by Justin Hileman
providing my own
```

So now the only bump in this path is that first banner line (which is where the `-q` option would help.)

And since you (kinda) asked, here's the full explanation for why that would be helpful:

* CakePHP 3 offers a starter skeleton that [includes Psysh](https://github.com/cakephp/app/blob/master/src/Shell/ConsoleShell.php#L58) to provide a command line environment where the entire Cake runtime environment is at your disposal inside the REPL. This turns out to be extremely handy, but you already know that. :wink: 
* I've [presented](https://github.com/beporter/CakePHP-EnvAwareness/tree/master/slides) a pretty stupidly simple way for Cake apps to maintain different configurations for different operating environments. This includes simple things like connecting to a local MySQL database when running inside vagrant, but connecting to RDS when running on EC2.
* I've also written a bunch of shell scripts that are meant to be used with a Cake app, including [one that looks up the database connection settings that Cake uses](https://github.com/loadsys/CakePHP-Shell-Scripts/blob/cake-2.x/db-credentials#L61), and [runs a database backup](https://github.com/loadsys/CakePHP-Shell-Scripts/blob/cake-2.x/db-backup) using those params (which ties into getting the "correct" settings for the current environment). This process is [currently pretty convoluted](https://github.com/loadsys/CakePHP-ConfigReadShell/issues/20) and involves [yet another Cake plugin](https://github.com/loadsys/CakePHP-ConfigReadShell) I had to write specifically to make this possible in Cake 2, but Psysh now offers hope of streamlining it like crazy by providing a really simple way to get access into the fully bootstrapped Cake environment, and just spit out serialized connection data (and already corrected for the local environment) to be consumed by external scripts. Psysh's banner line is the last thing standing in the way.

Now all that said, of course I can keep using `| tail -n+2` to strip the first line of output in the meantime, but rest assured, there is at least one use case for this. Although it's surely more work, potentially disabling the implicit return (`⏎`) during `--quiet` operation might also be worth considering, if the assumption is that, "I only want to see what the code **I** provide would output itself."

Thank again for your time. :+1:

**Comment by beporter** [comment]
So tl;dr: I guess:

The REPL has _a ton_ of use for being able to inject arbitrary code into a **bootstrapped** app environment and get values back that can be used externally-- provided that it's possible to get **just** the output you specify and not anything extraneous from Psysh itself.

**Comment by bobthecow** [comment]
This is fixed in c3a67e2d6f9c6a41809a9447ca9b395058071460 with the `-r` flag. It’ll be available in the next release :)
